File: crates/essence-wars-ui/src/lib/api/deckBuilder.ts
/**
 * Deck Builder API - Backend-agnostic wrappers.
 *
 * This module provides the public API for deck builder operations.
 * It delegates to the appropriate backend (Tauri or WASM) based on the platform.
 */

import { getGameBackend, getStorageBackend } from "./backends";
import type {
  BrowsableCard,
  CommanderDto,
  CustomDeck,
  CustomDeckInfo,
  DeckValidation,
  PlaystyleScore,
} from "./types";

// ===========================================================================
// Card Browsing
// ===========================================================================

/**
 * List all cards in the game.
 * @param faction Optional faction filter. If provided, returns cards from that faction plus neutral cards.
 * @returns Array of browsable cards
 */
export async function listAllCards(faction?: string): Promise<BrowsableCard[]> {
  return await getGameBackend().listAllCards(faction);
}

/**
 * List all commanders.
 * @returns Array of commander DTOs
 */
export async function listCommanders(): Promise<CommanderDto[]> {
  return await getGameBackend().listCommanders();
}

// ===========================================================================
// Custom Deck Management
// ===========================================================================

/**
 * List all custom decks (metadata only).
 * @returns Array of custom deck info
 */
export async function listCustomDecks(): Promise<CustomDeckInfo[]> {
  return await getStorageBackend().listCustomDecks();
}

/**
 * Load a custom deck by ID.
 * @param deckId The deck ID to load
 * @returns The full custom deck
 */
export async function loadCustomDeck(deckId: string): Promise<CustomDeck> {
  return await getStorageBackend().loadCustomDeck(deckId);
}

/**
 * Save a custom deck.
 * Creates a new deck or overwrites an existing one with the same ID.
 * @param deck The deck to save
 * @returns The deck ID on success
 */
export async function saveCustomDeck(deck: CustomDeck): Promise<string> {
  return await getStorageBackend().saveCustomDeck(deck);
}

/**
 * Delete a custom deck by ID.
 * @param deckId The deck ID to delete
 */
export async function deleteCustomDeck(deckId: string): Promise<void> {
  return await getStorageBackend().deleteCustomDeck(deckId);
}

// ===========================================================================
// Deck Validation & Analysis
// ===========================================================================

/**
 * Validate a custom deck configuration.
 * @param deck The deck to validate
 * @returns Validation result with errors and warnings
 */
export async function validateCustomDeck(deck: CustomDeck): Promise<DeckValidation> {
  return await getGameBackend().validateCustomDeck(deck);
}

/**
 * Calculate the playstyle for a deck based on its cards and commander.
 * @param cards The card IDs in the deck
 * @param commanderId The commander card ID
 * @returns The detected playstyle with scores
 */
export async function calculateDeckPlaystyle(
  cards: number[],
  commanderId: number
): Promise<PlaystyleScore> {
  return await getGameBackend().calculateDeckPlaystyle(cards, commanderId);
}

File: crates/essence-wars-ui/src-tauri/src/commands/deck_builder.rs
//! Deck builder Tauri commands.

use crate::state::{
    BrowsableCard, CommanderDto, CustomDeck, CustomDeckInfo, CustomDeckManager,
    DeckValidation, PlaystyleScore,
};
use tauri::State;

/// List all cards in the game (optionally filtered by faction).
///
/// If `faction` is provided, returns cards from that faction plus neutral cards.
/// If `faction` is None, returns all cards.
#[tauri::command]
pub fn list_all_cards(
    faction: Option<String>,
    deck_manager: State<'_, CustomDeckManager>,
) -> Vec<BrowsableCard> {
    match faction {
        Some(f) => deck_manager.list_cards_for_faction(&f),
        None => deck_manager.list_all_cards(),
    }
}

/// List all commanders.
#[tauri::command]
pub fn list_commanders(
    deck_manager: State<'_, CustomDeckManager>,
) -> Vec<CommanderDto> {
    deck_manager.list_commanders()
}

/// List all custom decks (metadata only).
#[tauri::command]
pub fn list_custom_decks(
    deck_manager: State<'_, CustomDeckManager>,
) -> Result<Vec<CustomDeckInfo>, String> {
    deck_manager.list_decks()
}

/// Load a custom deck by ID.
#[tauri::command]
pub fn load_custom_deck(
    deck_id: String,
    deck_manager: State<'_, CustomDeckManager>,
) -> Result<CustomDeck, String> {
    deck_manager.load_deck(&deck_id)
}

/// Save a custom deck.
///
/// Creates a new deck or overwrites an existing one with the same ID.
/// Returns the deck ID on success.
#[tauri::command]
pub fn save_custom_deck(
    deck: CustomDeck,
    deck_manager: State<'_, CustomDeckManager>,
) -> Result<String, String> {
    deck_manager.save_deck(&deck)
}

/// Delete a custom deck by ID.
#[tauri::command]
pub fn delete_custom_deck(
    deck_id: String,
    deck_manager: State<'_, CustomDeckManager>,
) -> Result<(), String> {
    deck_manager.delete_deck(&deck_id)
}

/// Validate a custom deck configuration.
///
/// Returns validation result with errors and warnings.
#[tauri::command]
pub fn validate_custom_deck(
    deck: CustomDeck,
    deck_manager: State<'_, CustomDeckManager>,
) -> DeckValidation {
    deck_manager.validate_deck(&deck)
}

/// Calculate the playstyle for a deck based on its cards and commander.
///
/// Returns the detected playstyle (Aggro, Control, Tempo, Midrange) with scores.
#[tauri::command]
pub fn calculate_deck_playstyle(
    cards: Vec<u16>,
    commander_id: u16,
    deck_manager: State<'_, CustomDeckManager>,
) -> PlaystyleScore {
    deck_manager.calculate_playstyle(&cards, commander_id)
}
